python/cudf/cudf/_lib/expressions.pyx

```diff
@@ -77,27 +77,20 @@ class TableReference(Enum):
 # restrictive at the moment.
 cdef class Literal(Expression):
     def __cinit__(self, value):
-        # TODO: Would love to find a better solution than unions for literals.
-        cdef int intval
-        cdef double doubleval
-
         if isinstance(value, int):
-            self.c_scalar_type = scalar_type_t.INT
-            intval = value
-            self.c_scalar.int_ptr = make_unique[numeric_scalar[int64_t]](
-                intval, True
-            )
+            self.c_scalar.reset(new numeric_scalar[int64_t](value, True))
```

Comment:
- Shouldn't every `new` have a corresponding `delete`? Note that if we need to call `delete`, it can be done in the `__dealloc__` method of this class
- Can we continue to use a unique/shared pointer so we don't have to worry about defining `__dealloc__` or calling `delete`?

Reply:
Oh I think I see what's going here, we're resetting the value of the unique pointer to a new `numeric_scalar`. But why do it this way instead of `make_unique`?

Reply:
`make_unique` created the `numeric_scalar` unique pointer. It can't be type casted and stored to generic base pointer. Hence this approach. I saw similar approach in another Cython file as well.
https://github.com/rapidsai/cudf/blob/branch-23.06/python/cudf/cudf/_lib/scalar.pyx#L250

Reply:
Thanks! I approved the PR, but still have a question. Why prefer:

```
c_scalar.reset(new ....)
```

versus:

```
c_scalar = make_unique(...)
```

?

Reply:
It would require releasing the unique_ptr of derived pointer anyway to store in  `unique_ptr<scalar>` class.
`unique_ptr[scalar] c_scalar = unique_ptr(static_cast<scalar*>(make_unique(...).release()));`

Hence using the new operator directly.

Reply:
Got it - thanks for clarifying! Definitely a TIL for me

>  Is 'releaseing the unique_ptr and typecasting' safer than using reset(new... ?

Do you mean is it safer from a Cython perspective? If so, the answer is that it should be no different than in C++. I would say whatever pattern you prefer in C++, we should use here as well.